src/nano_contracts/deserializer.ts

```diff
@@ -316,20 +307,27 @@ class Deserializer {
       // Address should be exactly 25 bytes long
       throw new Error('Address should be 25 bytes long');
     }
-    // First we get the 20 bytes of the address without the version byte and checksum
-    const addressBytes = buf.subarray(2, 22);
-    const address = helpersUtils.encodeAddress(addressBytes, this.network);
+    // The actual address bytes are the 25 bytes after the initial length
+    const addressBytes = buf.subarray(1);
+    // First we get the 20 bytes (hash) of the address without the version byte and checksum
+    const hashBytes = addressBytes.subarray(1, 21);
+    const address = helpersUtils.encodeAddress(hashBytes, this.network);
     address.validateAddress();
     const decoded = address.decode();
-    if (decoded[0] !== buf[1]) {
+    // We need to check that the metadata of the address received match the one we generated
+    // Check network version
+    if (decoded[0] !== addressBytes[0]) {
       throw new Error(
-        `Asked to deserialize an address with version byte ${buf[0]} but the network from the deserializer object has version byte ${decoded[0]}.`
+        `Asked to deserialize an address with version byte ${addressBytes[0]} but the network from the deserializer object has version byte ${decoded[0]}.`
       );
     }
-    if (decoded.subarray(21, 25).toString('hex') !== buf.subarray(22, 26).toString('hex')) {
+    // Check checksum bytes
+    if (!decoded.subarray(21, 25).equals(addressBytes.subarray(21, 25))) {
       // Checksum value generated does not match value from fullnode
+      const calcChecksum = decoded.subarray(21, 25).toString('hex');
+      const recvChecksum = addressBytes.subarray(21, 25).toString('hex');
```

Comment:
suggestion(non-blocking): The `calcChecksum` and `recvChecksum` could be declared before the `if`, so that they're used on both the conditional and the error message.